### PR TITLE
Add heater control functionality

### LINKED ISSUE #27: Heater control not implemented
the proccess in 5.2 of this https://www.silabs.com/documents/public/data-sheets/Si7021-A20.pdf

Arduino has it

### PR DESCRIPTION
**Pull request by tekktrik**
Resolves #27 by adding heater control in a single property which combines enabling and setting.  Uses `adafruit_register` so that dependency has been added.  Example usage in comment in the simpletest file.

**NOT TESTED** - Hardware isn't available for purchase right now so if anyone has it, please let me know!

### CODE REVIEW

**Review by anecdata — DISMISSED** [review]
Needs input validation? I was able to set heater to -1 and 99 and `"a"` without issue.

**Comment by tekktrik** [comment]
Ah, thank you!

**Review by jerryneedell — APPROVED** [review]
Tested with SI7021 breakout (non-stemma) on umfeathers3.
Ran with heater setting commented out
heater level =1
heater level = 5
All ran normally, higher heater level resulted in higher temperature measurements.
I did not monitor the current draw.

Looks good to me.

**Comment by jerryneedell** [comment]
> Needs input validation? I was able to set heater to -1 and 99 and `"a"` without issue.

Good catch!

**Comment by tekktrik** [comment]
Pushed argument validation for both type and range

**Comment by jerryneedell** [comment]
retested -- rejected 99 and 2.5 as desired
OK with 17 and 2
Note, setting it to 17 had a lessor impact not the temperature reading than setting it to 2....

**Comment by anecdata** [comment]
Should it explicitly set a heater level and enable state at init? Across a soft reload when I had left it at a high level unintentionally, I got crazy readings coming back in after some time had passed. Even the heater level didn't seem to get set right, like the sensor was malfunctioning.

Also, when I set it to `0`, it reads back at `16` and gets really hot. Maybe I fried it. But when I turn it down to 1, it cools off and reads fine.

```py
Adafruit CircuitPython 7.2.0 on 2022-02-24; Adafruit Feather ESP32-S2 TFT with ESP32S2
>>> import board
>>> import adafruit_si7021
>>> 
>>> sensor = adafruit_si7021.SI7021(board.I2C())
>>> 
>>> sensor.heater_level = 0
>>> sensor.heater_level
16
>>> sensor.heater_level = 1
>>> sensor.heater_level
1
```

File: adafruit_si7021.py [hunk 1]
```diff
@@ -177,6 +189,42 @@ def temperature(self):
...
+        """
+        if self._heater_enable_read:
+            return self._heater_level_read + 1
+        return 0
+
+    @heater_level.setter
+    def heater_level(self, level):
+        if not isinstance(level, int):
+            raise TypeError("Heater level must be int between 0 and 17, inclusive")
+        if not 0 <= level <= 17:
```
Comment (jerryneedell) [inline adafruit_si7021.py]: According to the data sheet the heater setting has is only 4 bits so it must be 0-15
so the max in the setting should be 16 with the level-1

**Review by jerryneedell — DISMISSED** [review]
I think the range of accepted values should be 0-16 not 17
according to the data sheet the heater control is 4 bits (0-15)
this explains the odd behavior at 17

**Review by jerryneedell — COMMENTED** [review]
Entering a 0 has a very odd response -- temperatures are very high.

**Comment by tekktrik** [comment]
It probably should be set explicitly on startup, I can add that

The control is 16 bits, but there's an entire separate bit in another register for enable (like level 0).  So I have it set where 0 interacts with that bit, and anything from 1-17 affects the the level as you mentioned.  It also helps alleviate the fact that according to the data sheet, setting those 4 bits to 0 still turned the heater on, so I tried to implement in a way that was more intuitive as a single property.

**Comment by jerryneedell** [comment]
> It probably should be set explicitly on startup, I can add that
> 
> The control is 16 bits, but there's an entire separate bit in another register for enable (like level 0). So I have it set where 0 interacts with that bit, and anything from 1-17 affects the the level as you mentioned. It also helps alleviate the fact that according to the data sheet, setting those 4 bits to 0 still turned the heater on, so I tried to implement in a way that was more intuitive as a single property.

I don't understand a value of 17 will write  b10000 -only the 4 low bits are used ad sets it to 0...


> Wait, it should be 0 - 17. Here's the intended behavior:
> 
> 0: Turn off the heater enable, don't worry about heater level 1 - 17: Subtract one from level, turn on heater enable, set level to calculated level

Yes, but 0 is not working that way.

**Comment by jerryneedell** [comment]
I repeated the test with a level of 0 and got the same result -- temperature reading went to > 50 C !!

Something is not correct.

**Comment by tekktrik** [comment]
Ah, whoops, thanks for writing it in binary, you're right it should be 0-16, one day I'll learn how to count xD

**Comment by jerryneedell** [comment]
I am concerned about the response to a level =0
I can look into it more later today, but something is very wrong there.
0 should turn it off, just like commenting out the line correct?

**Comment by tekktrik** [comment]
Wait, it should be 0 - 17.  Here's the intended behavior:

0:  Turn off the heater enable, don't worry about heater level
1 - 17: Subtract one from level, turn on heater enable, set level to calculated level

**Comment by tekktrik** [comment]
I wonder if the enable isn't working... if that's the case it could be that the level is getting set to -1...

**Comment by jerryneedell** [comment]
> Wait, it should be 0 - 17. Here's the intended behavior:
> 
> 0: Turn off the heater enable, don't worry about heater level 1 - 17: Subtract one from level, turn on heater enable, set level to calculated level

I do not agree 
a setting of 16 does not make any sense according to the data sheet.

0 is not working as intended

**Comment by jerryneedell** [comment]
I have to go for an hour or so -- I'll look into it more later today.
This may have to wait until you can get some hardware...

**Comment by tekktrik** [comment]
Sorry, I kept thinking b00001111 was 16, my bad.  In this case, it should be 0 <= level <= 16 which is a range of 17 numbers inclusive ( 0 for on/off, 1-16 for setting).

Still not sure why 0 doesn't work though...

**Comment by tekktrik** [comment]
My thought is that setting the level is actually doing something even if the enable is turned off... at level 0, subtracting 1 would give 0b11111111 in binary if I'm not mistaken, so the level is getting set at maximum

**Comment by anecdata** [comment]
maybe this for the enable register?

> For future compatibiliy, it is recommended that prior to a write operation, registers should be read. Then the values read from the RSVD bits should be written back unchanged during the write operation.

(maybe the register library takes care of that, I've never looked at it)

**Comment by tekktrik** [comment]
Good catch, I was thinking that `RWBit.__set__()` would bit mask for me, but I guess it can't unless I read first!

**Comment by tekktrik** [comment]
Looks like the classes I'm using might need some tweaks, I'll get on that and see if that can do the trick

**Comment by tekktrik** [comment]
Oh shoot, I had faulty logic - the heater was getting turned on every time it was set, even to 0!  Restructuring slightly but I think that was the main cause

**Comment by tekktrik** [comment]
Okay, reorganized things to basically borrow the `RWBit` and `RWBits` classes, but modify them to account for the fact that the read and write registers are different, this also removes the dependency on `adafruit_register`.

That aside, I think the main problem was so faulty logic I fixed in [7242f99](https://github.com/adafruit/Adafruit_CircuitPython_SI7021/pull/29/commits/7242f99460d0c13f9a33ebea6da8e7dd903f38af) that caused the heater to always be set on regardless

**Comment by jerryneedell** [comment]
Tested new version from 0 to 16 -- seems to work OK.
Also monitored current draw at each level -- appeared to be as expected

**Comment by tekktrik** [comment]
Awesome, thank you both!

**Review by jerryneedell — COMMENTED** [review]
I tested this as implemented and it now appears to be working as expected.
After thinking about it a bit more, I am wondering about the implementation.
Looking at the Arduino library, the heater enable and heater level are controlled independently and the heater level is set from 0-15 as stated in the data sheet .
For this implementation, you have combined the enable and level - using level 0 to disable the heater. The potential confusion is that now the heater levels are set 1 higher than the actual levels in the data sheet.
Is this the way we want to implement this? Should there be two properties. One for enable and one for level -- more like the Arduino library
https://github.com/adafruit/Adafruit_Si7021/blob/master/Adafruit_Si7021.cpp#L207
Both implementations work. I just wonder if this implementation add potential confusion.

**Comment by anecdata** [comment]
I'm comfortable with Jerry's review. The only other thing I noticed is that `from adafruit_bus_device.i2c_device import I2CDevice` is no longer needed?

I don't know all the pros and cons of using the register library vs. re-implementing a customized version here, but maybe we want to think about an addition or refactor to the register library for separate read & write registers so the feature can be shared with other libraries.

**Comment by tekktrik** [comment]
Hm, if the other library separates the two, I agree it may be worth doing the same.  I guess my only concern is that I wouldn't think a level of 0 for heater level would be anything, but per the datasheet it's the minimum.  Happy to change the implementation to whatever seems most sensible, that was just my initial take. 

The I2C Device is still used for grabbing the temperature and humidity data, so it should still be required.

I agree it may be worth pushing these changes to the register library.  The existing classes there are great for working with R/W or read only registers, bits, structs, etc. but they usually depend on the register to read and/or write to be the same.  I don't really know how common it is to separate them, but if it is, I agree add those changes would be worth it.

**Comment by anecdata** [comment]
> The I2C Device is still used for grabbing the temperature and humidity data, so it should still be required.

Of course, sorry, I just browser-searched the changes page, not the whole file.

**Comment by jerryneedell** [comment]
I think it is good to keep things consistent between the Arduino and CP libraries when possible.
I agree that it is a bit strange to have level 0 not be off, but it is documented in the data sheet. I guess they did not want waste a level. If you want it off, disable it ;-)
So, my suggestion is to separate the enable and level properties.
I have not used the heater on my devices and I'm not sure how, when it would be utilized. I t looks like the primary use is to help "dry out" the humidity sensor. I don't think it will be a problem if the enable and level are separated and I do think it is more consistent with the data sheet to do it that way.

**Comment by tekktrik** [comment]
Okay, I'll make those changes now

**Comment by jerryneedell** [comment]
> Okay, I'll make those changes now

That is fine with me, but I am only one reviewer -- feel free to solicit other comments.

**Comment by tekktrik** [comment]
@jerryneedell I think you're right and got some feedback that agrees, so let me know if it still works with new changes (altered the simpletest to expose it again in the commented out portion)

File: adafruit_si7021/__init__.py [hunk 2]
```diff
@@ -177,6 +187,47 @@ def temperature(self):
...
+        15            94.20
+        ============  =================
+
+        """
+        return self._heater_level
+
+    @heater_level.setter
+    def heater_level(self, level):
+        if not isinstance(level, int):
+            raise TypeError("Heater level must be int between 0 and 16, inclusive")
```
Comment (jerryneedell) [inline adafruit_si7021/__init__.py]: this should also be 0 and 15

Comment (tekktrik) [reply adafruit_si7021/__init__.py]: Good catch!

**Review by jerryneedell — APPROVED** [review]
This looks good to me. Thank you for making the changes.
I verified that with the enable set to false, setting the level has no impact on the temperature or current being drawn. That is reassuring.

**Comment by jerryneedell** [comment]
I don't know the best way to communicate this, but users should be alerted that they now have to install the entire adafruit_si7021 folder and not just adafruit_si7021.py
I guess it should go in the release notes and perhaps this should be a major release

**Comment by jerryneedell** [comment]
oops -- too late -- maybe just a note should be added to the Release notes.

**Comment by kattni** [comment]
@tekktrik @jerryneedell Releases are free. Tekktrik, please do another release, bump it to 4.0.0, and note in the release notes the major change. The guide will have to be updated as well, if there is one.

**Comment by kattni** [comment]
@tekktrik To be clear, this should be updated to not reference `.mpy` but have it be `adafruit_si7021` instead?
Screen Shot 2022-03-02 at 16 56 00

**Comment by tekktrik** [comment]
Ah, sorry about that, I can bump it up.  And yes!  The `.mpy` should be removed now, is that located in any of the repo documentation?

**Comment by kattni** [comment]
Don't know! Take a look please :)

**Comment by kattni** [comment]
Guide is updated.

**Comment by tekktrik** [comment]
Is this sufficient for a release note (along with standard text)?

The new heater control functionality also caused this library to grow into a collection of files (addition of `i2c_bits.py`) under a single folder.  This major revision update signifies this update.  It is sufficient to treat the folder as you would have the file (e.g., add it to your board by adding it your board).

**Comment by kattni** [comment]
Sure, that works!

**Comment by tekktrik** [comment]
Re-released!  Thanks y'all!

**Pull request merged**